[jit] Fix getAttribute

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29220 [quant][graphmode] Enable foldbn tests
* #30787 [jit] Add unsafeRemoveConstant for ClassType
* #29219 [jit] Expose class constant through `attr` and `setattr` in object
* #31012 [jit] Fix getConstant
* #29218 [jit] Preserve constant from ConcreteModuleType to ClassType
* **#31011 [jit] Fix getAttribute**

Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

jerryzh168 committed:
[jit] Fix getAttribute

Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

suo commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1432,7 +1432,7 @@ struct CAFFE2_API ClassType : public NamedType {
 
   TypePtr getAttribute(const std::string& name) const {
-    AT_ASSERT(attributeNames_.size() == attributeTypes_.size());
+    TORCH_CHECK(attributeNames_.size() == attributeTypes_.size());
```
This is an internal invariant check, it should be `TORCH_INTERNAL_ASSERT`

suo reviewed (changes requested):
As mentioned before, we should be defining:
```
TypePtr findAttribute()
TypePtr getAttribute()
bool hasAttribute()
```
As is, I think this PR is missing `findAttribute`.

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 3960b02:
* **1/3** broken upstream at merge base 6ab2d1b (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=6ab2d1b1a49de255da11d80772a0dd7ec9e9bae2&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **2/3** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=3960b0277c5e669b8345d0d223ded29388a34a87).

### 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3921511) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_simple_test (1/2)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=73003221))
```
Dec 12 07:19:48 Ran 2301 tests in 75.243s
Dec 12 07:19:48 
Dec 12 07:19:48 FAILED (failures=6, errors=23, skipped=83, expected failures=1)
Dec 12 07:19:48 
Dec 12 07:19:48 Generating XML reports...
Dec 12 07:19:48 Traceback (most recent call last):
Dec 12 07:19:48   File "test/run_test.py", line 456, in <module>
Dec 12 07:19:48     main()
Dec 12 07:19:48   File "test/run_test.py", line 449, in main
Dec 12 07:19:48     raise RuntimeError(message)
Dec 12 07:19:48 RuntimeError: test_jit_simple failed!
Dec 12 07:19:48 + cleanup
Dec 12 07:19:48 + retcode=1
Dec 12 07:19:48 + set +x
Dec 12 07:19:48 =================== sccache compilation log ===================
Dec 12 07:19:48 =========== If your build fails, please take a look at the log above for possible reasons ===========
Dec 12 07:19:48 Compile requests                56
Dec 12 07:19:48 Compile requests executed        0
Dec 12 07:19:48 Cache hits                       0
Dec 12 07:19:48 Cache misses                     0
Dec 12 07:19:48 Cache timeouts                   0
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3921513) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_legacy_test (2/2)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=73003989))
```
Dec 12 07:20:32 Ran 2301 tests in 114.830s
Dec 12 07:20:32 
Dec 12 07:20:32 FAILED (failures=6, errors=23, skipped=65, expected failures=1)
Dec 12 07:20:32 
Dec 12 07:20:32 Generating XML reports...
Dec 12 07:20:32 Traceback (most recent call last):
Dec 12 07:20:32   File "test/run_test.py", line 456, in <module>
Dec 12 07:20:32     main()
Dec 12 07:20:32   File "test/run_test.py", line 449, in main
Dec 12 07:20:32     raise RuntimeError(message)
Dec 12 07:20:32 RuntimeError: test_jit_legacy failed!
Dec 12 07:20:32 + cleanup
Dec 12 07:20:32 + retcode=1
Dec 12 07:20:32 + set +x
Dec 12 07:20:32 =================== sccache compilation log ===================
Dec 12 07:20:32 =========== If your build fails, please take a look at the log above for possible reasons ===========
Dec 12 07:20:32 Compile reretrieving test reports
```
### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_xla_linux_xenial_py3_6_clang7_build](https://dr.pytorch.org/build-details.html?build_id=72998759)



---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

suo commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1449,6 +1449,27 @@ struct CAFFE2_API ClassType : public NamedType {
     return attributeTypes_[pos];
   }
 
+  TypePtr getAttribute(const std::string& name) const {
```
Let's avoid this code duplication by having `getAttribute` call `findAttribute` internally.

suo commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1489,7 +1510,7 @@ struct CAFFE2_API ClassType : public NamedType {
         false,
         python_str(),
-        " does not have a field with the name '",
+        " does not have attribute field with the name '",
```
nit: this should be "does not have an attribute with the name"

suo reviewed (approved):
Accepting but please address the comments inline

suo reviewed (changes requested):
Actually, it looks like you will need to audit the existing callsites of `getAttribute` to make sure they're doing the right thing.

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 commented:
@suo please check the callsite

zdevito commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -660,7 +660,7 @@ class ShapePropagator {
         auto cls = node->input()->type()->expect<ClassType>();
         // propagate any type specializations encoded in the type of the class
-        node->output()->setType(cls->getAttribute(node->s(attr::name)));
+        node->output()->setType(cls->findAttribute(node->s(attr::name)));
```
this should be a getAttribute, setType expects a non-null type

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -609,7 +609,7 @@ struct slot_dict_impl {
 
   void setattr(const std::string& name, py::object value) {
-    const TypePtr& type = module_->type()->getAttribute(name);
+    const TypePtr& type = module_->type()->findAttribute(name);
```
getAttribute, toIValue expects a non-null type.

zdevito reviewed (commented):
Thanks for updating the API. I noted some places where find should probably be get. Ill like @suo do the approval when it is ready.

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

suo reviewed (approved):
Some additional callsites that need fixing:
- In object.h:33, there is a redundant `TORCH_CHECK` that can be removed.
- In python_sugared_value.cpp:220, there is a `getAttribute` where we check the `attrType` for nullity. That null check can be omitted, I'm not sure why it's there.
- In python_sugared_value.cpp:276, there is an if-condition that could be simplified with `findAttribute`

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/script/python_sugared_value.cpp:
```diff
@@ -273,7 +269,7 @@ std::shared_ptr<SugaredValue> ModuleValue::attr(
   const auto& selfType = selfType_->expect<ClassType>();
 
-  if (selfType->hasAttribute(field) && selfType->getAttribute(field)->is_module()) {
+  if (const auto& t = selfType->findAttribute(field) && t->is_module()) {
```
@suo is this what you meant?

suo replied:
yeah sure

jerryzh168 replied:
Unfortunately this does not compile..I think we still have to keep this check.

suo replied:
Okay let's drop it then. It's not particularly important.

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot retest this please

jerryzh168 committed:
Update on "[jit] Fix getAttribute"


Summary:
`getAttribute` is supposed to throw when there the attribute is not
found rather than return a `nullptr`.

Test Plan:
.

Reviewers:
suo

Subscribers:

Tasks:

Tags:

Differential Revision: [D18898417](https://our.internmc.facebook.com/intern/diff/D18898417)

[ghstack-poisoned]
